packages/loopring_v3/contracts/impl/libtransactions/OwnerChangeTransaction.sol

```diff
@@ -0,0 +1,189 @@
+        );
+
+        // Verify the signature if one is provided, otherwise fall back to an approved tx
+        if (auxData.signatureNewOwner.length > 0) {
+            require(txHash.verifySignature(accountTransfer.newOwner, auxData.signatureNewOwner), "INVALID_SIGNATURE");
+        } else {
+            require(S.approvedTx[accountTransfer.newOwner][txHash], "TX_NOT_APPROVED");
+            S.approvedTx[accountTransfer.newOwner][txHash] = false;
+        }
+
+        if (auxData.walletAddress == address(0)) {
+            // We allow the owner of the account to change when authorized by the current owner.
+            // Verify the signature if one is provided, otherwise fall back to an approved tx
+            if (auxData.signatureOldOwner.length > 0) {
+                require(txHash.verifySignature(accountTransfer.owner, auxData.signatureOldOwner), "INVALID_SIGNATURE");
+            } else {
+                require(S.approvedTx[accountTransfer.owner][txHash], "TX_NOT_APPROVED");
+                S.approvedTx[accountTransfer.owner][txHash] = false;
+            }
+        } else {
+            // If the account has a wallet, use it to recover the account
+            require(accountTransfer.walletHash != 0, "ACCOUNT_HAS_NO_WALLET");
+
+            // Calculate the wallet hash
+            bytes32 walletHash = EIP712.hashPacked(
+                S.DOMAIN_SEPARATOR,
+                keccak256(
+                    abi.encode(
+                        WALLET_TYPEHASH,
+                        auxData.walletAddress,
+                        auxData.walletDataHash
+                    )
+                )
+            );
+            // Hashes are stored using only 253 bits so the value fits inside a SNARK field element.
+            require((uint(walletHash) >> 3) == uint(accountTransfer.walletHash), "INVALID_WALLET_HASH");
+
+            // Check that the calldata contains the correct inputs for data coming from layer 2.
+            // This data is also padded with zeros to 32 bytes (at the MSB) inside the calldata.
+            // parameter 0: accountID
+            // parameter 1: nonce
+            // parameter 2: oldOwner
+            // parameter 3: newOwner
+            // parameter 4: walletDataHash
+            uint offset = 4;
+            require(auxData.walletCalldata.toUint24(29 + offset) == accountTransfer.accountID, "INVALID_WALLET_CALLDATA");
+            offset += 32;
+            require(auxData.walletCalldata.toUint32(28 + offset) == accountTransfer.nonce, "INVALID_WALLET_CALLDATA");
+            offset += 32;
+            require(auxData.walletCalldata.toAddress(12 + offset) == accountTransfer.owner, "INVALID_WALLET_CALLDATA");
+            offset += 32;
+            require(auxData.walletCalldata.toAddress(12 + offset) == accountTransfer.newOwner, "INVALID_WALLET_CALLDATA");
+            offset += 32;
+            require(auxData.walletCalldata.toBytes32(offset) == auxData.walletDataHash, "INVALID_WALLET_CALLDATA");
```

Comment:
From the protocol POV it's walletDataHash because it's the hash of the custom wallet data, from the wallet POV it's the hash of the wallet data. I can see how this is confusing. Changing the parameter name to walletDataHash makes the most sense then I guess to keep things consistent.